### ISSUE TITLE
FSI list of DateTimes taking too much space of output

### ISSUE THREAD

**Issue by Thorium** [fsharp/fslang-suggestions#1211]
I'm often doing FSI driven development and I evaluate a list of some complex objects and trying to read it through is a mess because any date-times take all the space:

![image](https://user-images.githubusercontent.com/229355/203338643-3340294c-1493-4ba5-8db1-245ddd5dc953.png)

Please, could we remove the `\r\n` line changes from this DateTime .ToString() presentation?

No-one cares the extra-properties like date-of-year taking all your screen space, and you have a wide-screen anyway, why not to utilize it?

**Comment by T-Gro** [comment]
Did you try using a custom printer via AddPrinter?
You can have your default printers in a common .fsx file and load it at top of your scripts.


```
open System
fsi.AddPrinter(fun (x : DateTime) -> x.ToString("dd.MM.yyyy hh:mm"))

let data = 
    [ {|Text = "dada" ; When = DateTime.UtcNow|}
      {|Text = "2112" ; When = DateTime.Today|}]
```

**Comment by baronfel** [comment]
Distribution/management of these snippets is hard, though - one more piece of evidence to me that some kind of FSI config (with machine, user, and local directory levels) might be nice to solve problems like this.

**Comment by T-Gro** [comment]
Thinking about convention-based, some well-known dot file names?

**Comment by Thorium** [comment]
No, I was watching an F# presentation of someone else and couldn't focus things that were not visible on his screen, so I came to create this suggestion.

**Comment by jbeeko** [comment]
> Distribution/management of these snippets is hard, though - one more piece of evidence to me that some kind of FSI config (with machine, user, and local directory levels) might be nice to solve problems like this.

There was talk of the FSI getting a directive the would let it load from GitHub. Would that not solve both this issue and others.

**Comment by jcmrva** [comment]
Or preferably any HTTP source.

**Comment by dsyme** [comment]
@T-Gro I do think we could just add a whole lot of stanard printers, or stop all those properties being dumped in the common case for .NET types.

**Comment by Thorium** [comment]
I kind of understand that part of the DateTime is culture (and time-zone?) so it's fair to say plain "01/01/2020" is not enough, because is it dd/mm or mm/dd.

**Comment by T-Gro** [comment]
> @T-Gro I do think we could just add a whole lot of stanard printers, or stop all those properties being dumped in the common case for .NET types.

Would it be acceptable to add new printers by default from a backwards compat standpoint?
It would change output for scripts already written.

If we can do this step, I would be for it - especially for beginners is better to have the visual clutter removed by default and not via additional step.

( Covering at least the ones frequently used as members of custom records:
- DateOnly in sortable format yyyy-MM-dd (to avoid cultural differences)
- TimeOnly  as HH:mm:ss
- DateTime same as   DateOnly +TimeOnly  
- TimeSpan  as TotalMilliseconds
- Uri as AbsoluteUri )

**Comment by Thorium** [comment]
This is somewhat related https://github.com/dotnet/fsharp/issues/17071

**Comment by T-Gro** [comment]
I am going to mark this as approved in principle, but would first like to get more feedback on any other types which are regularily encountered by scripting users and where there is hope on defining an agreed-upon fsi printer function.

I am not convinced about having the `byte array` printer default to UTF-8/ascii interpreted content out of the box, but it could come in as a function people can turn explicitely.

**Comment by Tarmil** [comment]
Sounds good, although I think TimeSpan should use ToString() rather than TotalMilliseconds, it's quite readable.

DateTimeOffset could be added too.
